### PR TITLE
feature/chained pubsub

### PR DESCRIPTION
**Pull request by rex-schilasky**
### Description
Added pubsub unit test to test publishing inside a subscriber callback.

### CODE REVIEW

File: ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp [hunk 1]
```diff
@@ -232,3 +232,69 @@ TEST(core_cpp_pubsub, TestPublisherIsSubscribedTiming)
...
+    std::cout << "Subscriber2 Receiving " << std::string(static_cast<const char*>(data->buf), data->size) << std::endl;
+    };
+  sub2.AddReceiveCallback(subscriber2_callback);
+
+  // Start publisher1 thread
+  std::thread pub1_thread(publisher1_thread);
+
+  // Wait until Publisher1 has sent all messages and Subscriber2 has received them
+  pub1_thread.join();
+  while (subscriber2_received_count < message_count)
```
Comment (KerstinKeller) [inline ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp:290]: This thread will loop forever, if not all messages were received.

File: ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp [hunk 2]
```diff
@@ -232,3 +232,69 @@ TEST(core_cpp_pubsub, TestPublisherIsSubscribedTiming)
...
+  // enable data loopback
+  eCAL::Util::EnableLoopback(true);
+
+  // Set up counters for sent and received messages
+  const int message_count = 10;
+  std::atomic<int> publisher1_sent_count(0);
+  std::atomic<int> subscriber2_received_count(0);
+
+  // Publisher1 in thread 1
+  auto publisher1_thread = [&]() {
```
Comment (KerstinKeller) [inline ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp:250]: Maybe rename variable (publisher function) and capture only neccesary variables.

File: ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp [hunk 3]
```diff
@@ -232,3 +232,69 @@ TEST(core_cpp_pubsub, TestPublisherIsSubscribedTiming)
...
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    }
+    };
+
+  // Publisher2
+  eCAL::CPublisher pub2("topic2");
+
+  // Subscriber1 with callback that triggers Publisher2
+  eCAL::CSubscriber sub1("topic1");
+  auto subscriber1_callback = [&](const char* /*topic_name*/, const eCAL::SReceiveCallbackData* data) {
```
Comment (KerstinKeller) [inline ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp:269]: capture only necessary info (publisher)

**Review by KerstinKeller — CHANGES_REQUESTED** [review]

File: ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp [hunk 4]
```diff
@@ -232,3 +230,71 @@ TEST(core_cpp_pubsub, TestPublisherIsSubscribedTiming)
...
+  sub2.AddReceiveCallback(subscriber2_callback);
+
+  // Start publisher1 thread
+  std::thread pub1_thread(publisher1_function);
+
+  // Wait until Publisher1 has sent all messages and Subscriber2 has received them
+  pub1_thread.join();
+  int max_message_loops(100);
+  while ((subscriber2_received_count < message_count) && (max_message_loops-- > 0))
+  {
```
Comment (KerstinKeller) [inline ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp:290]: Can you please explain the logic here?

Comment (rex-schilasky) [reply ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp:290]: That loop will run approx. 1 second at maximum (or if all messages sent by by pub2 are received).

Comment (rex-schilasky) [reply ecal/tests/cpp/pubsub_test/src/pubsub_connection_test.cpp:290]: Oh sorry, I see the problem. Fixed it.

**Review by KerstinKeller — APPROVED** [review]

**Pull request merged**